usd/src/sdf_parser/Joint.cc

```diff
@@ -0,0 +1,402 @@
+
+namespace sdf
+{
+// Inline bracke to help doxygen filtering.
+inline namespace SDF_VERSION_NAMESPACE {
+//
+namespace usd
+{
+  /// \brief Helper function for setting a USD joint's pose relative to the
+  /// joint's parent and child.
+  /// \param[in] _jointPrim The USD joint prim
+  /// \param[in] _joint The SDF representation of _jointPrim
+  /// \param[in] _parentModel The SDF model that is the parent of _joint
+  /// \return UsdErrors, which is a vector of UsdError objects. Each UsdError
+  /// includes an error code and message. An empty vector indicates no error
+  /// occurred when setting _joint's pose relative to _joint's parent and child
+  UsdErrors SetUSDJointPose(pxr::UsdPhysicsJoint &_jointPrim,
+      const sdf::Joint &_joint, const sdf::Model &_parentModel)
+  {
+    UsdErrors errors;
+
+    ignition::math::Pose3d parentToJoint;
+    if (_joint.ParentLinkName() == "world")
+    {
+      ignition::math::Pose3d modelToJoint;
+      auto poseErrors = usd::PoseWrtParent(_joint, modelToJoint);
+      if (!poseErrors.empty())
+        return poseErrors;
+
+      // it is assumed the _parentModel's parent is the world
+      ignition::math::Pose3d worldToModel;
+      poseErrors = usd::PoseWrtParent(_parentModel, worldToModel);
+      if (!poseErrors.empty())
+        return poseErrors;
+
+      parentToJoint = worldToModel * modelToJoint;
+    }
+    else
+    {
+      auto poseResolutionErrors =
+        _joint.SemanticPose().Resolve(parentToJoint, _joint.ParentLinkName());
+      if (!poseResolutionErrors.empty())
+      {
+        errors.push_back(UsdError(
+              sdf::Error(sdf::ErrorCode::POSE_RELATIVE_TO_INVALID,
+              "Unable to get the pose of joint [" + _joint.Name() +
+              "] w.r.t. its parent link [" + _joint.ParentLinkName() + "].")));
+        for (const auto &e : poseResolutionErrors)
+          errors.push_back(UsdError(e));
+        return errors;
+      }
+    }
+
+    ignition::math::Pose3d childToJoint;
+    auto poseResolutionErrors = _joint.SemanticPose().Resolve(childToJoint,
+        _joint.ChildLinkName());
+    if (!poseResolutionErrors.empty())
+    {
+      errors.push_back(UsdError(
+          sdf::Error(sdf::ErrorCode::POSE_RELATIVE_TO_INVALID,
+            "Unable to get the pose of joint [" + _joint.Name() +
+            "] w.r.t. its child [" + _joint.ChildLinkName() + "].")));
+      for (const auto &e : poseResolutionErrors)
+        errors.push_back(UsdError(e));
+      return errors;
+    }
+
+    auto parentRotation = pxr::GfQuatf(
+          parentToJoint.Rot().W(),
+          parentToJoint.Rot().X(),
+          parentToJoint.Rot().Y(),
+          parentToJoint.Rot().Z());
+
+    auto childRotation = pxr::GfQuatf(
+          childToJoint.Rot().W(),
+          childToJoint.Rot().X(),
+          childToJoint.Rot().Y(),
+          childToJoint.Rot().Z());
+
+    const auto axis = _joint.Axis();
+    if (axis && (axis->Xyz() == ignition::math::Vector3d::UnitY))
```

Comment:
this entire logic block seems strange to me. I'm sure there's a good reason, but it's not obvious. Can you add documentation to explain what this code is doing and why?

Reply:
I noticed that there's a `TODO` note here now. When doing original testing on a prototype a few months ago, @ahcorde found that joints along the y axis had to be handled differently in USD due to the USD axis conventions. I do not recall the specifics of the fix, but since there's a TODO note, I will mark this as resolved and we can revisit it later.